[jit] Implement more of of the nn.Module API

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28828 [jit] Implement more of of the nn.Module API**

This updates torch::script::Module to more closely match the behavior
of nn.Module. In particular, it implements the (optionally recurisive)
iterators that retrieve submodules, parameters, and buffers and makes
their names match the python versions.

This also removes the individual accessors for Parameter, Module, Buffer, etc.
and replaces them with a single `attr` function which is equivalent to
writing `a.foo` in Python (`setattr` emulates `a.foo = v`).
As we build out the user-facing API for TorchScript values this will end
up matching how an  attribute is accessed on general objects.

This PR preservers the python bindings for script::Module by emulating the
old API at the binding level. A followup will clean up the usage to more
directly match the C++ API.

Differential Revision: [D18197611](https://our.internmc.facebook.com/intern/diff/D18197611)

zdevito committed:
[jit] Implement more of of the nn.Module API

This updates torch::script::Module to more closely match the behavior
of nn.Module. In particular, it implements the (optionally recurisive)
iterators that retrieve submodules, parameters, and buffers and makes
their names match the python versions.

This also removes the individual accessors for Parameter, Module, Buffer, etc.
and replaces them with a single `attr` function which is equivalent to
writing `a.foo` in Python (`setattr` emulates `a.foo = v`).
As we build out the user-facing API for TorchScript values this will end
up matching how an  attribute is accessed on general objects.

This PR preservers the python bindings for script::Module by emulating the
old API at the binding level. A followup will clean up the usage to more
directly match the C++ API.

ZolotukhinM commented on torch/csrc/api/src/serialize/input-archive.cpp:
```diff
@@ -70,12 +56,11 @@ void InputArchive::read(
+  if (!module_.hasattr(key)) {
     return false;
   }
+  archive.module_ = module_.attr(key).toModule();
```
Wouldn't we crash here if the attribute is there, but it's not a module? In the original code we returned `false` in this case.

zdevito replied:
Yeah, I can make this more specific.

ZolotukhinM commented on torch/csrc/jit/script/module.h:
```diff
@@ -197,18 +235,21 @@ struct TORCH_API Module {
 
-  module_list get_modules() const;
+  module_list children() const; // direct modules
+  named_module_list named_children() const;
```
We should add a comment explaining the difference between `named_children` and `named_modules`. Ideally, add a comment for all these methods.

ZolotukhinM commented on torch/csrc/jit/script/module.h:
```diff
@@ -381,120 +391,237 @@ struct TORCH_API Module {
 // The template parameter allows us to use the same implementation for a list
 // that returns Module via template specialization of the operator* method.
-template <typename T>
+template <typename Policy>
```
The comment became stale as it still mentions `EntityType`.

ZolotukhinM commented on torch/csrc/jit/script/module.h:
```diff
@@ -381,120 +391,237 @@ struct TORCH_API Module {
-struct NameValue {
-  std::string name;
-  IValue value;
+struct Frame {
```
The name `Frame` is already overloaded in various contexts, are we sure we want to use it here as well?

zdevito replied:
Ill change it to something less ambiguous.

ZolotukhinM reviewed (commented):
I have two comments:
1) It seems that the old API for accessing attributes had checks for the type of the attribute (module/buffer/parameter), while the new API returns a generic one. We probably need to add these removed checks to the places where it was used to preserve the behavior (I commented in one of such sites, but there are probably more).
2) The code for recursive iterators is hard to understand. I know what it's supposed to be doing but it's difficult to follow it even with that knowledge. Can we please add some comments (and update the old comments)? Some classes like `Policy` et al would also benefit from brief comments.

zdevito committed:
Update on "[jit] Implement more of of the nn.Module API"

[jit] Implement more of of the nn.Module API

This updates torch::script::Module to more closely match the behavior
of nn.Module. In particular, it implements the (optionally recurisive)
iterators that retrieve submodules, parameters, and buffers and makes
their names match the python versions.

This also removes the individual accessors for Parameter, Module, Buffer, etc.
and replaces them with a single `attr` function which is equivalent to
writing `a.foo` in Python (`setattr` emulates `a.foo = v`).
As we build out the user-facing API for TorchScript values this will end
up matching how an  attribute is accessed on general objects.

This PR preservers the python bindings for script::Module by emulating the
old API at the binding level. A followup will clean up the usage to more
directly match the C++ API.

gh-metadata: pytorch pytorch 28828 gh/zdevito/130/head

zdevito committed:
Update on "[jit] Implement more of of the nn.Module API"

[jit] Implement more of of the nn.Module API

This updates torch::script::Module to more closely match the behavior
of nn.Module. In particular, it implements the (optionally recurisive)
iterators that retrieve submodules, parameters, and buffers and makes
their names match the python versions.

This also removes the individual accessors for Parameter, Module, Buffer, etc.
and replaces them with a single `attr` function which is equivalent to
writing `a.foo` in Python (`setattr` emulates `a.foo = v`).
As we build out the user-facing API for TorchScript values this will end
up matching how an  attribute is accessed on general objects.

This PR preservers the python bindings for script::Module by emulating the
old API at the binding level. A followup will clean up the usage to more
directly match the C++ API.

gh-metadata: pytorch pytorch 28828 gh/zdevito/130/head

zdevito committed:
Update on "[jit] Implement more of of the nn.Module API"

[jit] Implement more of of the nn.Module API

This updates torch::script::Module to more closely match the behavior
of nn.Module. In particular, it implements the (optionally recurisive)
iterators that retrieve submodules, parameters, and buffers and makes
their names match the python versions.

This also removes the individual accessors for Parameter, Module, Buffer, etc.
and replaces them with a single `attr` function which is equivalent to
writing `a.foo` in Python (`setattr` emulates `a.foo = v`).
As we build out the user-facing API for TorchScript values this will end
up matching how an  attribute is accessed on general objects.

This PR preservers the python bindings for script::Module by emulating the
old API at the binding level. A followup will clean up the usage to more
directly match the C++ API.

gh-metadata: pytorch pytorch 28828 gh/zdevito/130/head

ZolotukhinM commented on torch/csrc/jit/script/module.h:
```diff
@@ -381,120 +393,291 @@ struct TORCH_API Module {
+struct TORCH_API AttributePolicy {
+  using value_type = IValue;
+  static value_type create(
+      const std::vector<detail::SlotCursor>& frames,
```
Nit: s/frames/cursors/

ZolotukhinM commented on torch/csrc/jit/script/module.h:
```diff
@@ -381,120 +393,291 @@ struct TORCH_API Module {
+      IValue v) {
+    std::string name;
+    if (frames.size() == 1) {
+      name = (frames.back().i_ == -1) ? "" : nameFragment(frames.back());
```
Do we really have to special case `size==1`? It looks like it can be handled just fine by the loop below.

zdevito replied:
It's to avoid the overhead of allocating a ostringstream and copying the string twice in the very common non-recursive case.

ZolotukhinM commented on torch/csrc/jit/script/module.h:
```diff
@@ -381,120 +393,291 @@ struct TORCH_API Module {
+        if (i > 0) {
+          ss << ".";
+        }
+        ss << nameFragment(frames[i]);
```
What if `frames[i].i_ == -1`? Should we assert that it's not the case?

zdevito replied:
The getAttributeName is going to assert in that case, so I didn't add another one. It's not a bug because only the top-level frame can have this.

ZolotukhinM commented on torch/csrc/jit/script/module.h:
```diff
@@ -381,120 +393,291 @@ struct TORCH_API Module {
-           (type_ && module_.entity_type(i_) != *type_)) {
-      ++i_;
+  // return_module() is a corner case where instead of returning a submodule
+  // of root, we are return root itself, because we are iterating modules(),
```
Typo: "are return"

ZolotukhinM commented on torch/csrc/jit/script/module.h:
```diff
@@ -381,120 +393,291 @@ struct TORCH_API Module {
+  // return_module() is a corner case where instead of returning a submodule
+  // of root, we are return root itself, because we are iterating modules(),
+  // which contains the root module itself.
+  // It is represented with a single Frame whose index is -1.
```
Nit: since Frame was renamed to SlotCursor, we need to update all its references accordingly.

ZolotukhinM commented on torch/csrc/jit/script/module.h:
```diff
@@ -381,120 +393,291 @@ struct TORCH_API Module {
+    }
+    // the last traversal action advanced beyond the number of slots in the
+    // module so continue the iteration in the parent.
+    if (top().i_ >= int64_t(top().module_.num_slots())) {
```
Should we do it *while* we're beyond the number of slots instead of *if*? I.e. shall we pop back until we reach a valid position or do we intentionally want to pop back only once?

zdevito replied:
That would be correct too, but I went with this way to make the components easier to understand: next() does 1 step, while_not_valid is responsible for repeating.

ZolotukhinM reviewed (approved):
Overall looks good, please find some comments inline.

zdevito commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -433,6 +433,128 @@ bool ivalue_tags_match(const Module& lhs, const Module& rhs) {
+  slot_dict_impl<Policy>(self.module_object()).setattr(name, std::move(value));
+}
+
+static py::object get_generic(Module& self, const std::string& name) {
```
My next patch is going to remove all of those and replace them with `attr`, so I didn't bother with it here. All of these are private methods in our ScriptModule implementation that are already guarded.

zdevito commented on torch/csrc/jit/script/module.cpp:
```diff
@@ -138,19 +137,11 @@ void Module::to_impl(
     const c10::optional<at::Device>& device,
     const c10::optional<at::ScalarType>& dtype,
     bool non_blocking) {
-  // First call `to()` on every child module.
```
yep, a nice side effect of having recursive things built in.

zdevito committed:
Update on "[jit] Implement more of of the nn.Module API"

[jit] Implement more of of the nn.Module API

This updates torch::script::Module to more closely match the behavior
of nn.Module. In particular, it implements the (optionally recurisive)
iterators that retrieve submodules, parameters, and buffers and makes
their names match the python versions.

This also removes the individual accessors for Parameter, Module, Buffer, etc.
and replaces them with a single `attr` function which is equivalent to
writing `a.foo` in Python (`setattr` emulates `a.foo = v`).
As we build out the user-facing API for TorchScript values this will end
up matching how an  attribute is accessed on general objects.

This PR preservers the python bindings for script::Module by emulating the
old API at the binding level. A followup will clean up the usage to more
directly match the C++ API.

gh-metadata: pytorch pytorch 28828 gh/zdevito/130/head

zdevito committed:
Update on "[jit] Implement more of of the nn.Module API"

[jit] Implement more of of the nn.Module API

This updates torch::script::Module to more closely match the behavior
of nn.Module. In particular, it implements the (optionally recurisive)
iterators that retrieve submodules, parameters, and buffers and makes
their names match the python versions.

This also removes the individual accessors for Parameter, Module, Buffer, etc.
and replaces them with a single `attr` function which is equivalent to
writing `a.foo` in Python (`setattr` emulates `a.foo = v`).
As we build out the user-facing API for TorchScript values this will end
up matching how an  attribute is accessed on general objects.

This PR preservers the python bindings for script::Module by emulating the
old API at the binding level. A followup will clean up the usage to more
directly match the C++ API.

gh-metadata: pytorch pytorch 28828 gh/zdevito/130/head

kostmo commented:
## CircleCI build failures summary
As of commit 2c07826:
* 0/2 flaky
* 2/2 failures introduced in this PR



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=2c078262f59a0857085f7819c5436cb5558cbf61).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

zdevito committed:
Update on "[jit] Implement more of of the nn.Module API"

[jit] Implement more of of the nn.Module API

This updates torch::script::Module to more closely match the behavior
of nn.Module. In particular, it implements the (optionally recurisive)
iterators that retrieve submodules, parameters, and buffers and makes
their names match the python versions.

This also removes the individual accessors for Parameter, Module, Buffer, etc.
and replaces them with a single `attr` function which is equivalent to
writing `a.foo` in Python (`setattr` emulates `a.foo = v`).
As we build out the user-facing API for TorchScript values this will end
up matching how an  attribute is accessed on general objects.

This PR preservers the python bindings for script::Module by emulating the
old API at the binding level. A followup will clean up the usage to more
directly match the C++ API.

gh-metadata: pytorch pytorch 28828 gh/zdevito/130/head

zdevito committed:
Update on "[jit] Implement more of of the nn.Module API"

[jit] Implement more of of the nn.Module API

This updates torch::script::Module to more closely match the behavior
of nn.Module. In particular, it implements the (optionally recurisive)
iterators that retrieve submodules, parameters, and buffers and makes
their names match the python versions.

This also removes the individual accessors for Parameter, Module, Buffer, etc.
and replaces them with a single `attr` function which is equivalent to
writing `a.foo` in Python (`setattr` emulates `a.foo = v`).
As we build out the user-facing API for TorchScript values this will end
up matching how an  attribute is accessed on general objects.

This PR preservers the python bindings for script::Module by emulating the
old API at the binding level. A followup will clean up the usage to more
directly match the C++ API.

gh-metadata: pytorch pytorch 28828 gh/zdevito/130/head

zdevito committed:
Update on "[jit] Implement more of of the nn.Module API"

[jit] Implement more of of the nn.Module API

This updates torch::script::Module to more closely match the behavior
of nn.Module. In particular, it implements the (optionally recurisive)
iterators that retrieve submodules, parameters, and buffers and makes
their names match the python versions.

This also removes the individual accessors for Parameter, Module, Buffer, etc.
and replaces them with a single `attr` function which is equivalent to
writing `a.foo` in Python (`setattr` emulates `a.foo = v`).
As we build out the user-facing API for TorchScript values this will end
up matching how an  attribute is accessed on general objects.

This PR preservers the python bindings for script::Module by emulating the
old API at the binding level. A followup will clean up the usage to more
directly match the C++ API.

gh-metadata: pytorch pytorch 28828 gh/zdevito/130/head

zdevito committed:
Update on "[jit] Implement more of of the nn.Module API"

[jit] Implement more of of the nn.Module API

This updates torch::script::Module to more closely match the behavior
of nn.Module. In particular, it implements the (optionally recurisive)
iterators that retrieve submodules, parameters, and buffers and makes
their names match the python versions.

This also removes the individual accessors for Parameter, Module, Buffer, etc.
and replaces them with a single `attr` function which is equivalent to
writing `a.foo` in Python (`setattr` emulates `a.foo = v`).
As we build out the user-facing API for TorchScript values this will end
up matching how an  attribute is accessed on general objects.

This PR preservers the python bindings for script::Module by emulating the
old API at the binding level. A followup will clean up the usage to more
directly match the C++ API.

gh-metadata: pytorch pytorch 28828 gh/zdevito/130/head

zdevito committed:
Update on "[jit] Implement more of of the nn.Module API"

[jit] Implement more of of the nn.Module API

This updates torch::script::Module to more closely match the behavior
of nn.Module. In particular, it implements the (optionally recurisive)
iterators that retrieve submodules, parameters, and buffers and makes
their names match the python versions.

This also removes the individual accessors for Parameter, Module, Buffer, etc.
and replaces them with a single `attr` function which is equivalent to
writing `a.foo` in Python (`setattr` emulates `a.foo = v`).
As we build out the user-facing API for TorchScript values this will end
up matching how an  attribute is accessed on general objects.

This PR preservers the python bindings for script::Module by emulating the
old API at the binding level. A followup will clean up the usage to more
directly match the C++ API.

gh-metadata: pytorch pytorch 28828 gh/zdevito/130/head

zdevito committed:
Update on "[jit] Implement more of of the nn.Module API"

[jit] Implement more of of the nn.Module API

This updates torch::script::Module to more closely match the behavior
of nn.Module. In particular, it implements the (optionally recurisive)
iterators that retrieve submodules, parameters, and buffers and makes
their names match the python versions.

This also removes the individual accessors for Parameter, Module, Buffer, etc.
and replaces them with a single `attr` function which is equivalent to
writing `a.foo` in Python (`setattr` emulates `a.foo = v`).
As we build out the user-facing API for TorchScript values this will end
up matching how an  attribute is accessed on general objects.

This PR preservers the python bindings for script::Module by emulating the
old API at the binding level. A followup will clean up the usage to more
directly match the C++ API.

gh-metadata: pytorch pytorch 28828 gh/zdevito/130/head

zdevito committed:
Update on "[jit] Implement more of of the nn.Module API"

[jit] Implement more of of the nn.Module API

This updates torch::script::Module to more closely match the behavior
of nn.Module. In particular, it implements the (optionally recurisive)
iterators that retrieve submodules, parameters, and buffers and makes
their names match the python versions.

This also removes the individual accessors for Parameter, Module, Buffer, etc.
and replaces them with a single `attr` function which is equivalent to
writing `a.foo` in Python (`setattr` emulates `a.foo = v`).
As we build out the user-facing API for TorchScript values this will end
up matching how an  attribute is accessed on general objects.

This PR preservers the python bindings for script::Module by emulating the
old API at the binding level. A followup will clean up the usage to more
directly match the C++ API.

gh-metadata: pytorch pytorch 28828 gh/zdevito/130/head

pietern commented:
@zdevito This raced with #29208 and broke master.

Working on a fix instead of reverting because I think it's an easy fix.

pietern commented:
Not an easy fix after all.

Reverting #29208 now.
